File: cpp/atbash-cipher/atbash_cipher.cpp
#include "atbash_cipher.h"

using namespace std;

string atbash::encode(string s) {
  string res = "";
  int num_chars = 0;
  for (size_t i = 0; i < s.length(); i++) {
    char prev_char = tolower(s[i]);
    if ((prev_char >= 'a' && prev_char <= 'z') || (prev_char >= '0' && prev_char <= '9')) {
      if (num_chars != 0 && num_chars % 5 == 0) {
        res += ' ';
      }
      num_chars++;
      if (prev_char >= '0' && prev_char <= '9') {
        res += prev_char;
      } else {
        res += char('z' - (prev_char - 'a'));
      }
    }
  }
  return res;
}

string atbash::decode(string s) {
  string res = "";
  for (size_t i = 0; i < s.length(); i++) {
    if (s[i] >= 'a' && s[i] <= 'z') {
        res += char('z' - (s[i] - 'a'));
    } else if (s[i] >= '0' && s[i] <= '9') {
        res += s[i];
    }
  }
  return res;
}

File: csharp/atbash-cipher/AtbashCipher.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System.Linq;

public static class AtbashCipher
{
    public static string Encode(string plainValue)
    {
        var cleaned = Regex.Replace(plainValue.ToLower(), @"[,. ]", "");
        var blocks = new List<string>();
        var currBlock = "";

        for (var i = 0; i < cleaned.Length; i++)
        {
            var c = cleaned[i];
            currBlock += Char.IsLetter(c) ? Translate(c) : c;
            // add space where necessary
            if (i % 5 == 4 && i != cleaned.Length - 1)
            {
                blocks.Add(currBlock);
                currBlock = "";
            }
        }
        blocks.Add(currBlock);
        return String.Join(" ", blocks);
    }

    public static string Decode(string encodedValue) =>
        String.Join("", encodedValue.Where(c => c != ' ')
            .Select(c => Char.IsLetter(c) ? Translate(c) : c)
        );

    private static char Translate(char c) => (char)('z' - (Char.ToUpper(c) - 'A'));
}
